File: 0238-product-of-array-except-self/0238-product-of-array-except-self.cpp
class Solution {
public:
    vector<int> productExceptSelf(vector<int>& nums) {
        vector<long long>left(nums.size()),right(nums.size());
        vector<int>ans(nums.size());
        int r=nums.size()-1;
        left[0]=nums[0];
        right[r]=nums[r];
        for(int i=1;i<nums.size();i++){
            left[i]=left[i-1]*nums[i];
            r--;
            right[r]=right[r+1]*nums[r];
        }
        for(int i=0;i<nums.size();i++){
            ans[i]=1;
            if(i-1>=0){
                ans[i]=left[i-1];
            }
            if(i+1<nums.size()){
                ans[i]*=right[i+1];
            }
        }
        return ans;
    }
};

File: 0238-product-of-array-except-self/0238-product-of-array-except-self.java
import java.util.Arrays;

public class Solution {
    public int[] productExceptSelf(int[] nums) {
        int n = nums.length;
        int[] left = new int[n];
        int[] right = new int[n];
        int[] ans = new int[n];
        
        left[0] = nums[0];
        right[n - 1] = nums[n - 1];
        
        for (int i = 1; i < n; i++) {
            left[i] = left[i - 1] * nums[i];
            right[n - 1 - i] = right[n - i] * nums[n - 1 - i];
        }
        
        ans[0] = right[1];
        ans[n - 1] = left[n - 2];
        
        for (int i = 1; i < n - 1; i++) {
            ans[i] = left[i - 1] * right[i + 1];
        }
        
        return ans;
    }

    public static void main(String[] args) {
        Solution solution = new Solution();
        int[] nums = {1, 2, 3, 4};
        int[] result = solution.productExceptSelf(nums);
        System.out.println(Arrays.toString(result));
    }
}
